function-images/tests/chained-function-serving/tracing/tracing.go

```diff
@@ -0,0 +1,80 @@
+package tracing
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/trace/zipkin"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv"
+)
+
+func initTracer(tp *sdktrace.TracerProvider) func() {
+	otel.SetTracerProvider(tp)
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{}))
+	return func() {
+		_ = tp.Shutdown(context.Background())
+	}
+}
+
+func makeZipkinExporter(url string, logger *log.Logger) *zipkin.Exporter {
+	exporter, err := zipkin.NewRawExporter(
+		url,
+		zipkin.WithLogger(logger),
+		zipkin.WithSDKOptions(sdktrace.WithSampler(sdktrace.AlwaysSample())),
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return exporter
+}
+
+func InitBasicTracer(url string, serviceName string) func() {
+	var logger = log.New(os.Stderr, "tracer-log", log.Ldate|log.Ltime|log.Llongfile)
+	hostname, ok := os.LookupEnv("HOSTNAME")
+	if !ok {
+		hostname = ""
+	}
+	exporter := makeZipkinExporter(url, logger)
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sdktrace.AlwaysSample()),
+		sdktrace.WithSyncer(exporter),
+		sdktrace.WithResource(resource.NewWithAttributes(
+			semconv.ServiceNameKey.String(fmt.Sprintf("%v@%v", serviceName, hostname)),
+			attribute.Int64("ID", 1),
+		)),
+	)
+	f := initTracer(tp)
+	return f
+}
+
+func InitCustomTracer(url string, traceRate float64, logger *log.Logger, attr ...attribute.KeyValue) func() {
+	exporter := makeZipkinExporter(url, logger)
+	var sampler sdktrace.Sampler
+	if traceRate >= 1 {
+		sampler = sdktrace.AlwaysSample()
+	} else {
+		sampler = sdktrace.ParentBased(sdktrace.TraceIDRatioBased(traceRate))
+	}
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sampler),
+		sdktrace.WithSyncer(exporter),
+		sdktrace.WithResource(resource.NewWithAttributes(
+			attr...,
+		)),
+	)
+	f := initTracer(tp)
+	return f
+}
+
```

Comment:
I suggest the following refactoring. create a Span type with 2 pointer receivers: StartSpan() and EndSpan()